apps/webservice/src/app/[workspaceSlug]/systems/[systemSlug]/environments/TargetFilterUniquenessIndicator.tsx

```diff
@@ -0,0 +1,169 @@
+import type { Target } from "@ctrlplane/db/schema";
+import type { Node } from "reactflow";
+import React, { useMemo } from "react";
+import { IconAlertTriangle, IconCircleFilled } from "@tabler/icons-react";
+import { useQueries } from "@tanstack/react-query";
+
+import { cn } from "@ctrlplane/ui";
+import {
+  Tooltip,
+  TooltipContent,
+  TooltipProvider,
+  TooltipTrigger,
+} from "@ctrlplane/ui/tooltip";
+
+import { api } from "~/trpc/react";
+
+type UniqueFilterResult = {
+  isUnique: boolean;
+  overlaps: {
+    nodeId: string;
+    overlappingTargets: Target[];
+  }[];
+};
+
+export const useTargetFilterUniqueness = (
+  nodes: Node[],
+  workspaceId: string,
+): UniqueFilterResult | null => {
+  const utils = api.useUtils();
+  const filteredNodes = nodes.filter((node) => node.data.targetFilter);
+
+  const queries = useQueries({
+    queries: filteredNodes.map((node) => ({
+      queryKey: ["target", workspaceId, node.data.targetFilter, node.id],
+      queryFn: () =>
+        utils.target.byWorkspaceId.list.fetch({
+          workspaceId,
+          filter: node.data.targetFilter,
+        }),
+      enabled: Boolean(workspaceId),
+    })),
+  });
+
+  const isLoading = queries.some((query) => query.isLoading);
+  const isError = queries.some((query) => query.isError);
+
+  const result = useMemo<UniqueFilterResult | null>(() => {
+    if (!workspaceId || filteredNodes.length === 0 || isLoading) return null;
+    if (isError) return { isUnique: false, overlaps: [] };
+
+    const results = queries.map((query, index) => ({
+      nodeId: filteredNodes[index]?.id ?? "",
+      targets: query.data?.items ?? [],
+    }));
+
+    const targetToNodesMap = results.reduce<Record<string, Set<string>>>(
+      (acc, { nodeId, targets }) => {
+        targets.forEach((target) => {
+          acc[target.id] ??= new Set<string>();
+          acc[target.id]?.add(nodeId);
+        });
+        return acc;
+      },
+      {},
+    );
+
+    const duplicateTargetIds = Object.entries(targetToNodesMap)
+      .filter(([_, nodeIds]) => nodeIds.size > 1)
+      .map(([targetId]) => targetId);
+
+    if (duplicateTargetIds.length === 0)
+      return { isUnique: true, overlaps: [] };
+
+    const overlaps = results
+      .map(({ nodeId, targets }) => ({
+        nodeId,
+        overlappingTargets: targets.filter((target) =>
+          duplicateTargetIds.includes(target.id),
+        ),
+      }))
+      .filter(({ overlappingTargets }) => overlappingTargets.length > 0);
+
+    return {
+      isUnique: overlaps.length === 0,
+      overlaps,
+    };
+  }, [workspaceId, filteredNodes, queries, isLoading, isError]);
+
+  return result;
+};
+
+export const TargetFilterUniquenessIndicator: React.FC<{
+  nodes: Node[];
+  workspaceId: string;
+}> = ({ nodes, workspaceId }) => {
+  const result = useTargetFilterUniqueness(nodes, workspaceId);
+
+  const tooltipContent = !result ? (
```

Comment:
if you have to assign react to a variable, you are probably doing something wrong. Nest it, or make it its own component